#! /bin/sh -e
## 30gnumeric.dpatch by upstream Zhe Su
##   Fix Bug#280660: scim kills data in gnumeric while moving cursor
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fix gnumeric data loss problem while moving cursor

if [ $# -lt 1 ]; then
    echo "`basename $0`: script expects -patch|-unpatch as argument" >&2
    exit 1
fi

[ -f debian/patches/00patch-opts ] && . debian/patches/00patch-opts
patch_opts="${patch_opts:--f --no-backup-if-mismatch} ${2:+-d $2}"

case "$1" in
    -patch) patch -p1 ${patch_opts} < $0;;
    -unpatch) patch -R -p1 ${patch_opts} < $0;;
    *)
        echo "`basename $0`: script expects -patch|-unpatch as argument" >&2
        exit 1;;
esac

exit 0

@DPATCH@
diff -u -r1.116.2.1 -r1.116.2.3
--- scim-1.0.1.orig/extras/gtk2_immodule/gtkimcontextscim.cpp	11 Sep 2004 13:18:37 -0000	1.116.2.1
+++ scim-1.0.1/extras/gtk2_immodule/gtkimcontextscim.cpp	15 Nov 2004 14:41:50 -0000	1.116.2.3
@@ -23,7 +23,7 @@
  * Free Software Foundation, Inc., 59 Temple Place, Suite 330,
  * Boston, MA  02111-1307  USA
  *
- * $Id: gtkimcontextscim.cpp,v 1.116.2.1 2004/09/11 13:18:37 suzhe Exp $
+ * $Id: gtkimcontextscim.cpp,v 1.116.2.3 2004/11/15 14:41:50 suzhe Exp $
  */
 
 #define Uses_SCIM_DEBUG
@@ -266,6 +266,7 @@
 static GdkColor              _active_text;
 
 static gint                  _snooper_id                 = 0;
+static bool                  _snooper_installed          = false;
 
 static int                   _input_context_count        = 0;
 
@@ -490,6 +491,11 @@
 {
     SCIM_DEBUG_FRONTEND(1) << "gtk_im_context_scim_focus_in...\n";
 
+    if (!_snooper_installed) {
+        _snooper_id = gtk_key_snooper_install ((GtkKeySnoopFunc)gtk_scim_key_snooper, NULL);
+        _snooper_installed = true;
+    }
+
     GtkIMContextSCIM *context_scim = GTK_IM_CONTEXT_SCIM (context);
 
     if (context_scim && context_scim->impl) {
@@ -503,6 +509,12 @@
 gtk_im_context_scim_focus_out (GtkIMContext *context)
 {
     SCIM_DEBUG_FRONTEND(1) << "gtk_im_context_scim_focus_out...\n";
+
+    if (_snooper_installed) {
+        gtk_key_snooper_remove (_snooper_id);
+        _snooper_installed = false;
+    }
+
     GtkIMContextSCIM *context_scim = GTK_IM_CONTEXT_SCIM (context);
 
     if (context_scim && context_scim->impl && context_scim == _focused_ic) {
@@ -1607,6 +1619,7 @@
 
     /* XXX:This is not recommended way!! */
     _snooper_id = gtk_key_snooper_install ((GtkKeySnoopFunc)gtk_scim_key_snooper, NULL);
+    _snooper_installed = true;
 }
 
 static void
@@ -1763,13 +1776,18 @@
 {
     SCIM_DEBUG_FRONTEND(1) << "slot_hide_preedit_string...\n";
     if (_focused_ic && _focused_ic->impl && _focused_ic->impl->id == id) {
-        _focused_ic->impl->preedit_string = WideString ();
-        _focused_ic->impl->preedit_caret = 0;
-        _focused_ic->impl->preedit_attrlist.clear ();
-        if (_focused_ic->impl->use_preedit)
-            g_signal_emit_by_name(_focused_ic, "preedit_changed");
-        else
+        bool emit=false;
+        if (_focused_ic->impl->preedit_string.length ()) {
+            _focused_ic->impl->preedit_string = WideString ();
+            _focused_ic->impl->preedit_caret = 0;
+            _focused_ic->impl->preedit_attrlist.clear ();
+            emit = true;
+        }
+        if (_focused_ic->impl->use_preedit) {
+            if (emit) g_signal_emit_by_name(_focused_ic, "preedit_changed");
+        } else {
             panel_req_hide_preedit_string (_focused_ic);
+        }
     }
 }
 
@@ -1794,11 +1812,13 @@
 {
     SCIM_DEBUG_FRONTEND(1) << "slot_update_preedit_caret...\n";
     if (_focused_ic && _focused_ic->impl && _focused_ic->impl->id == id) {
-        _focused_ic->impl->preedit_caret = caret;
-        if (_focused_ic->impl->use_preedit)
-            g_signal_emit_by_name(_focused_ic, "preedit_changed");
-        else
-            panel_req_update_preedit_caret (_focused_ic, caret);
+        if (_focused_ic->impl->preedit_caret != caret) {
+            _focused_ic->impl->preedit_caret = caret;
+            if (_focused_ic->impl->use_preedit)
+                g_signal_emit_by_name(_focused_ic, "preedit_changed");
+            else
+                panel_req_update_preedit_caret (_focused_ic, caret);
+        }
     }
 }
 
@@ -1811,6 +1831,7 @@
     if (_focused_ic && _focused_ic->impl && _focused_ic->impl->id == id) {
         _focused_ic->impl->preedit_string   = str;
         _focused_ic->impl->preedit_attrlist = attrs;
+        _focused_ic->impl->preedit_caret    = str.length ();
         if (_focused_ic->impl->use_preedit)
             g_signal_emit_by_name(_focused_ic, "preedit_changed");
         else
